### PR TITLE
Implement remote cache storing whole query

### PR DESCRIPTION
**Pull request by andrewma2**
## What changes are proposed in this pull request?
Implementing a remote cache version of the [hackathon cache](https://docs.google.com/document/d/1X4Oiry-aWWfb6EwPh6QyeM0cK7NEO3HxBkWk-mWWRd8/edit#heading=h.g7ieoocb671f). We store results in a remote Redis server.

## How is this tested?
Tested using comparison test at go/m3dev

Deployed to `dev-aws-us-west-1`. Monitored performance metrics on [dashboard](https://grafana.cloud.databricks.com/d/jfb0fkenz/m3coordinator-cache-monitoring?orgId=1&from=now-1h&to=now). Observed caching rate similar to local hackathon cache rate

Was able to turn off Redis caching (checked logs to verify that caching was/wasn't used) by exec-ing the following command (in the pod):
`curl -X POST -F "toggle={true/false}" http://localhost:7201/api/v1/redis/toggle`

True means use the cache, False means don't

#### Code Review
For information about the code review process, e.g. how to find a reviewer or how to ping non-responsive reviewers, check the contents of [go/code-review](http://go/code-review) Confluence page.

#### Approvals
Other than the mandatory approvers enforced by the OWNER file framework (http://go/owners), this PR
requires at least one approval from another engineer.

#### [NEW] Shiproom

**Platform & Compute Fabric**:
Changes should be tracked by an approved "material change." Multiple PRs may be tracked by a single material change.

- [ ] Change modifies code owned or released by a Platform or Compute Fabric team
  - [ ] A "material change" covering this PR exists in http://go/engshiproom: `<CHANGE_ID>`

See http://go/platshiproomwiki for instructions and use http://go/lightcm-template to evaluate risk. Ask questions in #platform-shiproom.

**Runtime changes**:
- [ ] Change targets a runtime maintenance release (i.e., targets a maintenance `dbr-branch-x.x` branch or has a maintenance `dbr-branch-x.x` label)
  - [ ] Change is **NOT** a “material change”
  - [ ] Change **IS** a “material change” of low / medium risk
  - [ ] Change **IS** a “material change” of high risk needing Shiproom review

Please refer Runtime Shiproom Wiki: http://go/runtimeshiproomwiki

#### Security implications
_This section is intended for the reviewers of this PR_
Please, make sure you consider the content of "What are the responsibilities of code reviewers?" section of [go/pr-security-review](http://go/pr-security-review)

### CODE REVIEW

File: src/query/storage/cache/redis_cache.go [hunk 1]
```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
...
 }
+
+// Given a fetch query, converts it into a key for Redis
+func keyEncode(query *storage.FetchQuery) string {
+	res := make([]string, len(query.TagMatchers))
+	for i, m := range query.TagMatchers {
+		res[i] = m.String()
+	}
+	// Sort to guarantee we have the same order every time
+	sort.Strings(res)
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:63]: Curious, why this need to be order?

File: src/query/storage/cache/redis_cache.go [hunk 2]
```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
...
+
+	label_key := strings.Join(res, ";")
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:65]: looks like you define a schema for the key, can you put the schema example in the comment?

File: src/query/storage/cache/redis_cache.go [hunk 3]
```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
...
+	res, err := result.PromResult.Marshal()
+	if err != nil {
+		return "", err
+	}
+	return string(res), nil
+}
+
+// Given a string, decodes it into a result (or nil if failed)
+//
+// Sets the metadata to a blank as we don't use it
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:91]: What is in the metadata?

File: src/query/storage/cache/redis_cache.go [hunk 4]
```diff
@@ -1,26 +1,37 @@
...
+	ExpirationTime string = "300"
+	// Blank result used to denote an empty PromResult
+	EmptyResult string = "{}"
 	// Minimum number of connections pools to Redis to keep open
 	MinPools int = 10
 	// Time to wait before creating a new Redis connection pool (in ms)
 	CreateAfterTime time.Duration = 100 * time.Millisecond
 )
 
+type RedisCache struct {
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:29]: Can we put a dynamic config to enable it or disable this cache? This will be easier for your to write an integration tests.

File: src/query/storage/cache/redis_cache.go [hunk 5]
```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
...
+	// Consume as bytes to avoid having to casting to string later on (avoids duplicate memory)
+	responses := make([][]byte, len(entries))
+	if err := cache.client.Do(radix.Cmd(&responses, "MGET", keys...)); err != nil {
+		cache.logger.Error("Failed to execute Redis batch get", zap.Error(err))
+		return results
+	}
+	for i, r := range responses {
+		// If the response is nil (not EmptyResult), then it doesn't exist
+		if len(r) == 0 {
+			results[i] = nil
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:129]: for some result == nil, are you going to send this to m3db?

File: src/query/storage/cache/redis_cache.go [hunk 6]
```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
...
+	}
+
+	res := cache.Get([]*storage.FetchQuery{align_q})
+	if res[0] == nil {
+		promRes, err := st.FetchProm(ctx, q, fetchOptions)
+		if err == nil {
+			cache.Set([]*storage.FetchQuery{align_q}, []*storage.PromResult{&promRes})
+		}
+
+		cache.logger.Info("cache miss", zap.String("key", keyEncode(align_q)), zap.Int("size", len(promRes.PromResult.Timeseries)))
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:213]: beside logging, can we also add metric ?

File: src/query/storage/cache/redis_cache.go [hunk 7]
```diff
@@ -41,3 +52,167 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
...
+	keys := make([]string, len(entries))
+	results := make([]*storage.PromResult, len(entries))
+
+	for i, b := range entries {
+		keys[i] = keyEncode(b)
+	}
+
+	// Consume as bytes to avoid having to casting to string later on (avoids duplicate memory)
+	responses := make([][]byte, len(entries))
+	if err := cache.client.Do(radix.Cmd(&responses, "MGET", keys...)); err != nil {
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:122]: can we add err handling on the error? 
1. retry with backoff
2. emit metrics to read from client with err logging
3. err then fallback to m3 read, do crash coordinator.

**Review by davidyuanfs — COMMENTED** [review]
Some initial comments first, will take a look the unit test

File: src/query/storage/cache/redis_cache.go (hunk 1 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:63]: Just in case that two queries that ask for the same things but they come in a different order. For example, `http_requests{code=404, method="GET"}` and `http_requests{method="GET", code=404}` are the same but may not have the labels come in the same order so the key might look different without it

File: src/query/storage/cache/redis_cache.go (hunk 5 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:129]: Yeah so it will mark this result as nil and then later in the `WindowGetOrFetch` function, it will send it to M3DB if the result is nil

File: src/query/storage/cache/redis_cache.go (hunk 6 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:213]: Yeah, I have [another PR](https://github.com/andrewma2/m3/pull/4) for adding metrics that depends on this one, just split it into two for readability

File: src/query/storage/cache/redis_cache.go (hunk 3 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:91]: It's [this struct](https://github.com/m3db/m3/blob/5765747a4abc4bb5a9b2a38a647982f734a72ea5/src/query/block/meta.go#L112-L144). I omitted it since there isn't a function to serialize this part of the struct, only the actual results and the metadata doesn't seem to affect any accuracy.

File: src/query/storage/cache/redis_cache.go (hunk 7 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:124]: So to clarify on step 3, right now the code should fallback to reading from M3DB when an error is encountered, but I'm unsure as to what do crash coordinator means

Comment (davidyuanfs) [reply src/query/storage/cache/redis_cache.go:124]: sorry, clarify more on the comment

File: src/query/storage/cache/redis_cache.go (hunk 6 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:264]: Update: Added metrics to this PR

File: src/query/storage/cache/redis_cache.go (hunk 4 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:30]: Added a new HTTP endpoint to disable cache

File: src/query/storage/cache/redis_cache.go (hunk 7 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:173]: Will update in a future PR as this one is already getting a bit big

File: src/query/storage/cache/redis_cache.go [hunk 8]
```diff
@@ -1,29 +1,88 @@
...
 
-type RedisCache struct {
-	client       radix.Client
-	redisAddress string
-	logger       *zap.Logger
-}
-
 const (
+	// Time that keys expire after (in seconds)
+	ExpirationTime string = "300"
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:21]: I thought we will use LRU cache? It is OK if you want to update in the next pr.

File: src/query/storage/cache/redis_cache.go [hunk 9]
```diff
@@ -35,9 +94,179 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
...
+	fetchOptions *storage.FetchOptions,
+	q *storage.FetchQuery,
+	cache *RedisCache,
+) (storage.PromResult, error) {
+	if cache == nil || !EnableCache {
+		return st.FetchProm(ctx, q, fetchOptions)
+	}
+	query_range := int64(q.End.Sub(q.Start).Seconds())
+	// Align to the current minute
+	align_end := int64(q.End.Unix()/60) * 60
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache.go:255]: minor: put comment why we need to align with current minute?

File: src/query/storage/cache/redis_cache_test.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+	"github.com/stretchr/testify/require"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+)
+
+var (
+	cache RedisCache
+)
+
+func TestMain(m *testing.M) {
```
Comment (davidyuanfs) [inline src/query/storage/cache/redis_cache_test.go:26]: Can we add unit test to cover WindowGetOrFetch method with some corner cases?

**Review by davidyuanfs — APPROVED** [review]
overall looks good to me, worth to add more unit test to cover the corner cases return from cache vs storage.

File: src/query/storage/cache/redis_cache.go (hunk 8 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:21]: Yeah in the sliding window version, it'll have no expiration and allow Redis to use LRU

File: src/query/storage/cache/redis_cache_test.go (hunk 10 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache_test.go:26]: So these unit tests are run locally without making any connection to an M3DB instance and making a connection to one might be a bit tricky, can maybe replace with a mock storage, will work to add more corner cases

Comment (andrewma2) [reply src/query/storage/cache/redis_cache_test.go:26]: Update: found a way to test it without M3DB by basing it off some predefined unit tests in M3 already

File: src/query/storage/cache/redis_cache.go [hunk 11]
```diff
@@ -35,9 +94,186 @@ func NewRedisCache(redisAddress string, logger *zap.Logger) *RedisCache {
...
+	if err := result.PromResult.Unmarshal(val); err != nil {
+		return nil, err
+	}
+	result.Metadata = block.NewResultMetadata()
+	return &result, nil
+}
+
+func (cache *RedisCache) FlushAll() {
+	var response string
+	cache.client.Do(radix.Cmd(&response, "FLUSHALL"))
```
Comment (jnyi) [inline src/query/storage/cache/redis_cache.go:164]: Would you add some comments about this method, what data we get from `flushall`?

File: src/query/storage/cache/redis_cache_test.go [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/generated/proto/prompb"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage"
+	radix "github.com/mediocregopher/radix/v3"
```
Comment (jnyi) [inline src/query/storage/cache/redis_cache_test.go:16]: Is there a particular reason we use a personal developers' redis golang reop instead of the offical one: https://github.com/go-redis/redis

**Review by jnyi — COMMENTED** [review]
Hi @andrewma2 ,

Sorry for the delayed review due to oncall, but looks good to me overall, left small comments.

File: src/query/storage/cache/redis_cache_test.go (hunk 12 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache_test.go:16]: Yeah I was basing it off of the ratelimit code which also uses this package, but switching to go-redis shouldn't be that hard of a change

File: src/query/storage/cache/redis_cache.go (hunk 11 above)
Comment (andrewma2) [reply src/query/storage/cache/redis_cache.go:164]: Yeah `flushall` just clears the cache and I just wrote it to for utility in a local unit test, will update with comments
